calicoctl/resourcemgr/resourcemgr.go

```diff
@@ -56,78 +83,238 @@ type ResourceActionCommand func(*client.Client, unversioned.Resource) (unversion
 type resourceHelper struct {
-	typeMetadata      unversioned.TypeMetadata
+	resource          runtime.Object
+	listResource      ResourceListObject
 	resourceType      reflect.Type
 	tableHeadings     []string
 	tableHeadingsWide []string
 	headingsMap       map[string]string
 	isList            bool
-	apply             ResourceActionCommand
+	isNamespaced      bool
 	create            ResourceActionCommand
 	update            ResourceActionCommand
 	delete            ResourceActionCommand
+	get               ResourceActionCommand
 	list              ResourceActionCommand
 }
 
 func (r resourceHelper) String() string {
-	return fmt.Sprintf("Resource(%s %s)", r.typeMetadata.Kind, r.typeMetadata.APIVersion)
+	if !r.isList {
+		return fmt.Sprintf("Resource(%s %s)", r.resource.GetObjectKind(), r.resource.GetObjectKind().GroupVersionKind())
+
+	}
+	return fmt.Sprintf("Resource(%s %s)", r.listResource.GetObjectKind(), r.listResource.GetListMeta().GetResourceVersion())
 }
 
-// Store a resourceHelper for each resource unversioned.TypeMetadata.
-var helpers map[unversioned.TypeMetadata]resourceHelper
+// Store a resourceHelper for each resource.
+var helpers map[schema.GroupVersionKind]resourceHelper
+var kindToRes = make(map[string]runtime.Object)
 
-func registerResource(res unversioned.Resource, resList unversioned.Resource,
+func registerResource(res runtime.Object, resList runtime.Object, isNamespaced bool, names []string,
 	tableHeadings []string, tableHeadingsWide []string, headingsMap map[string]string,
-	apply, create, update, delete, list ResourceActionCommand) {
+	create, update, delete, get, list ResourceActionCommand) {
 
 	if helpers == nil {
-		helpers = make(map[unversioned.TypeMetadata]resourceHelper)
+		helpers = make(map[schema.GroupVersionKind]resourceHelper)
 	}
 
-	tmd := res.GetTypeMetadata()
 	rh := resourceHelper{
-		typeMetadata:      tmd,
+		resource:          res,
 		resourceType:      reflect.ValueOf(res).Elem().Type(),
 		tableHeadings:     tableHeadings,
 		tableHeadingsWide: tableHeadingsWide,
 		headingsMap:       headingsMap,
 		isList:            false,
-		apply:             apply,
+		isNamespaced:      isNamespaced,
 		create:            create,
 		update:            update,
 		delete:            delete,
+		get:               get,
 		list:              list,
 	}
-	helpers[tmd] = rh
+	helpers[res.GetObjectKind().GroupVersionKind()] = rh
 
-	tmd = resList.GetTypeMetadata()
 	rh = resourceHelper{
-		typeMetadata:      tmd,
+		listResource:      resList.(ResourceListObject),
 		resourceType:      reflect.ValueOf(resList).Elem().Type(),
 		tableHeadings:     tableHeadings,
 		tableHeadingsWide: tableHeadingsWide,
 		headingsMap:       headingsMap,
 		isList:            true,
 	}
-	helpers[tmd] = rh
+	helpers[resList.GetObjectKind().GroupVersionKind()] = rh
+
+	for _, v := range names {
+		kindToRes[v] = res
+	}
+}
+
+func (rh resourceHelper) GetObjectType() reflect.Type {
+	return rh.resourceType
+}
+
```

Comment:
Grrr.  Github lost some comments I'd made on this... Oh well, here we go again....